### ISSUE TITLE
Additional Definitions/Theorems for Words

### ISSUE THREAD

**Issue by avekens** [metamath/set.mm#599]
In my Mathbox, I provided some additional definitions and many additional theorems about Words (df-word), see section 19.23.4 "Additional definitions and more theorems about Words". I would propose to take them over into the main part of set.mm. 

Doing this, the sections  5.6.9  "Words over a set" and 5.6.10  "Longer string literals" should be restructured. I would propose the following structure (lifting up the section 5.6.9 by one level, incrementing the number of the following sections by 1):

5.7.  Words over a set
5.7.1 Definitions and basic theorems
5.7.2 Last symbol of a word
5.7.3 Singleton words
5.7.4 Concatenations of words
5.7.5 Concatenations with singleton words
5.7.6 Subwords
5.7.7 Subwords of subwords
5.7.8 Subwords of concatenations
5.7.9 Splicing words (substring replacement)
5.7.10 Reversing words
5.7.11 Cyclic shifts of words
5.7.12 Longer string literals

What is your oppinion about this? I think Words are used very often in other Mathboxes, so we can collect the common "knowlegde" in the main part.

**Comment by nmegill** [comment]
I am basically OK with this but would like to hear another opinion.  Anyone else - @digama0 ? - is this appropriate to make into a major section at the level of "Elementary integer functions"?

**Comment by avekens** [comment]
PS: If changing section "5.6.9 Words over a set" to a major section "5.7. Words over a set" is not acceptable, we can use the 4th level of headings:
5.6.9 . Words over a set
5.6.9.1 Definitions and basic theorems
5.6.9.2 Last symbol of a word
5.6.9.3 Singleton words
...

**Comment by nmegill** [comment]
@avekens I asked Gerard to give an opinion, and he recommends it and has some additional remarks.  I am happy to go along with his suggestions if you are, and you are welcome to make the changes to the main set.mm.

Dear Norm,
 
1: In my personal opinion, Alexander’s proposition to create in the main part of set.mm, inside « PART 5 REAL AND COMPLEX NUMBERS »  a new paragraph « 5.7 Words over a set », mainly built with the current section ‘5.6.9 Words over a set", followed by transferring the current section « 19.23.4  Additional definitions and more theorems about words » and also containing at the end the current section « 5.6.10 Longer strings literals » is most adequate because clearly Words are a basic concept having many applications in differents parts of mathematics.
2: This would mean that the currents paragraphs (and sections) from 5.7 to 5.11 would be renumbered from  5.8 to 5.12.
3: Concerning the new paragraph « 5.7 Words over a set », the list of its sections as proposed by Alexander seems also correct.
4; I would also advocate for the following additions of very basic and useful results in the main part of set.mm:
(i)   Transfer of Alexander’s section « 19.23.3.25 The floor (greatest integer) function - extension «  in the current section « 5.6.1 The floor (greatest integer) function »;
(ii)  Transfer of Alexander’s section « 19.23.4.26 The modulo (remainder) operation - extension » in the current section  « 5.6.2 The modulo (remainder) operation »;
(iii) Transfer of Alexander’s section « 19.23.4.27 The « hash » (finite set size) function - extension »  (only from hashimarn 28740 to hashss 28747, that do not use set theory), and also of theorems  hashresfn 24330 and dmhashres 24331 from Thierry’s section « 19.3.5.7 The hash (finite set size) function - misc additions », in the current  section  « 5.6.8  The « hash » (finite set size) function ».
5: Finally, I would also add in the new paragraph « 5.7 Words over a set » the current section « 19.3.18.1 Operations on words » from Thierry’s Mathbox
6: And maybe also add in the current section « 5.7.2 Signum (sgn or sign) function «  (to become section 5.8.2) Thierrys current section « 19.3.17 Signum (sgn or sign) function - misc additions » from theorem sgncl 25188 to theorem eluzmn 25203 (that do not use monois theory).

Très amicalement.
Gérard Lang

**Comment by david-a-wheeler** [comment]
I would like to see these somewhere in the main body of set.mm.  The more general-use theorems we have in the main body, the better.  They'll be easier to use,  we'll have more consistency, and theorems will generally be easier to prove (especially since mmj2 will automatically find them if you try to use the relevant expression in a proof).

I don't care so much how they're ordered, though I would like it to have *some* logical order.

**Comment by avekens** [comment]
OK, thanks to everybody for the (positive) feedback. I will still wait for a while (if there are more comments/opinions) before I start with steps 1. to 3. of Gérard's list. Doing this, I probably need to move some of the auxiliary theorems which are used in the proof also to the main part. 

Only after finishing this, I'll look for steps 4. to 6., but basically I agree that these steps should be done.

**Comment by digama0** [comment]
I think the theorems themselves could use some review. Some comments:

* `wrdlen1`: This can be simplified using `risset`, and the length assumption can be weakened so it becomes redundant with `wrdsymb1`.
* `iswrd0i`: nonstandard naming (maybe put a `fz` in there instead)
* `wrdsymb0`: awkward hypotheses (split into two theorems, use a single conjunct for all hypotheses)
* `2wrdeq`: duplicate of `wrdeq0`. Perhaps this theorem should be called `eqwrdfv` by analogy to `eqfnfv`?
* `wrdlenge2n0`: a bit over-specific (surely we want the hypothesis to be weakened to length != 0 or > 0), I would not look for a theorem like this in the library
* `wrd11s1`: use 3an, flip biconditional order
* `wrdlen2`: shouldn't this be biconditional?
* `swrd0len0`: hypothesis should be weakened to `N < ( # ` W )` (maybe debatable)
* `s2eq2s1eq`: the right side seems a bit useless. Shouldn't it just be `A = C /\ B = D`?
* `wwlktovfun`, `wwlktovinj`, `wwlktovsur` , `wwlktovbij0`: nonstandard naming - the names for these predicates are `f`, `f1`, `fo`, `f1o`
* `wrdlenccats1lenm1`: this can be generalized to ` ( # ` P ) = ( ( # ` ( P concat <" A "> ) ) - 1 ) )`. Also, the preferred variable used for words seems to be inconsistent, you should pick a convention and stick to it
* `elovmpt2wrd`, `elovmptnn0wrd`: seems a bit over-specific, there are a few options for generalization
* `swrdltnd`, `swrdnd`: avoid multiple conjunct groups
* `swrd0`: what's up with the remark?
* `addlenrevswrd`: seems like the addends would make more sense the other way around
* `disjxwrd`: s/disjunct/disjoint/. Also it has an easy proof by `invdisj`
* `swrd0swrd`: could be proven as a special case of `swrdswrd`
* `swrdccat3a0`: generalizes to `( (/) concat W ) = W`
* `swrdccatin1d`: deduction form should split up all the assumptions into different hypotheses
* `add2t1`: sounds like it involves multiplication. Maybe one of `add1add1` or `add1p1` or `p1p1`?
* `ccat2s1cl`, ...: is there a reason to not use the analogous theorems on `<" X Y ">`? The proofs are also much longer than if you just rewrote to this, maybe you weren't aware of them?
* `ccatw2s1ass`: As above, I would want some specific advantage of being specific to singleton words. This should be ` ( ( W concat <" X "> ) concat <" Y "> ) = ( W concat <" X Y "> )`, and there shouldn't be any other theorems about `( ( W concat <" X "> ) concat <" Y "> )`
* `ccatw2s1cl`: reducible
* `ccat2s1fvw`: I am concerned why this isn't a four line proof
* `df-csh`: should be lowercase - `cyclShift`
* `df-lswrd`: should be lowercase - `lastS`. Is `last` taken?
* `cshwoor`: could be merged with the previous theorem with ``-. N e. ( 0 ... ( # ` W ) )``. We should try to minimize how much we discuss out of bounds behavior.
* Also, couldn't `df-csh` be extended to ZZ by modular arithmetic? That would give it some more regular behavior, although you probably want to only talk about how it factors into `substr` when the input is between 0 and ``(# ` W)``
* `lswrdn0`: Really not a fan of the empty set assumption. This should just say that it is in `V`. Also, the name segment for this function is inconsistent. Is it `ls` or `lswrd`?
* `lswrd1`: I think it would be nicer to just say ``( lastS ` <" A "> ) = A``. In fact you should have these theorems for all the explicit tuples, and a theorem about ``( lastS ` ( S concat T ) ) = ( lastS ` T )``
* `lstccats1fst`: this is bizzarely stated. The RHS should just be ``( P ` 0 )``. In fact it should be generalized to ``( lastS ` ( S concat <" A "> ) ) = A``, as a consequence of the other theorems I mentioned
* `swrdtrcfvl`: the length of this proof is surprising. It should be `swrd0fvls` + `swrd0len` + algebra
* `2cshwcom`: should follow from `2cshwmod` + algebra
* `cshwsame`: could be stated with `A` instead of ``( W ` 0 )`` in the hypothesis. Also, I think this is a good place to introduce the idiom `( ( 0 ..^ L ) X. { A } )` for the word "repeat A, L times"
* `cshwsdisj`, ...: split up the deduction hypothesis
* `cshwssize`: it is called `hash` not `size`
* `cshwsexa`: requires no assumptions
* `lstwrdne`: duplicate of `lswrdcl` (also, more naming inconsistency)
* `lstwrdne0`: I don't think we need it. The two assumptions can be merged to ``( # ` W ) e. NN`, but that's just another way to say `W =/= (/)` which is the preferred idiom.
* `2swrd1eqwrdeq`, `2swrd2eqwrdeq`: again, avoid `<" A "> = <" B ">`
* `ccat2s1lst`: reducible
* `ccats1swrdeq`: you could state this with ``( lastS ` U )`` instead of an existential
* `ccats1rev`: These hypotheses are pretty crazy. If you want sufficient conditions for the implication, I would suggest ``( lastS ` U ) = S``, but then the conclusion is reducible to just ``U = ( W concat <" ( lastS ` U ) "> )``.

**Comment by avekens** [comment]
Thanks, Mario, for the comments/hints. I will revise the theorems accordingly in my mathbox first. After finishing this (and maybe another review) I can start moving the theorems to the main part.

**Comment by avekens** [comment]
I just created a pull request with the my latest changes, mainly according to Mario's review remarks. Most of the issues should be solved now. There are only a few, which I will leave in my Mathbox for the moment when I move the additional material to the main part (the theorems which I want to move are marked with *RC* - Release Candidates). This concerns despecially the definition (and corresponding theorems) about cyclical shifts: here I want to investigate in Mario's suggestion to generalize the definition - this could have an impact on many theorems...

I als added a proposal for the section header comment of the section "Words over a set" (see subsection "Additional definitions and more theorems about Words"), including conventions for the usage of class variables.

Details for each of Mario's remarks will follow.

**Comment by avekens** [comment]
Here the details for each of Mario's remarks:
* wrdlen1: This can be simplified using risset, and the length assumption can be weakened so it becomes redundant with wrdsymb1.
  * proof shortened using wrdsymb1
  * will remain in AV's Mathbox
  * wrdsymb1: preferred variable used for words adjusted P->W
* iswrd0i: nonstandard naming (maybe put a fz in there instead)
  * renamed to ffz0iswrd
* wrdsymb0: awkward hypotheses (split into two theorems, use a single conjunct for all hypotheses)
  * to be clarified/analysed
  * will remain in AV's Mathbox
* 2wrdeq: duplicate of wrdeq0. Perhaps this theorem should be called eqwrdfv by analogy to eqfnfv?
  * already replaced and removed!
  * wrdeq0: renamed to eqwrd 
* wrdlenge2n0: a bit over-specific (surely we want the hypothesis to be weakened to length != 0 or > 0), I would not look for a theorem like this in the library
  * will remain in AV's Mathbox
* wrdl1s1: use 3an, flip biconditional order
  * OK, changed accordingly
* wrdlen2: shouldn't this be biconditional?
  * wrdlen2 renamed to wrdlen2i
  * new wrdlen2 biconditional now
* swrd0len0: hypothesis should be weakened to N < ( # W )` (maybe debatable)
  * will remain in AV's Mathbox
* s2eq2s1eq: the right side seems a bit useless. Shouldn't it just be A = C /\ B = D?
  * s2eq2s1eq preserved: usage in this form, easier to prove (because of df-s2).
  * s2eq2seq added with right side A = C /\ B = D
* wwlktovfun, wwlktovinj, wwlktovsur , wwlktovbij0: nonstandard naming - the names for these predicates are f, f1, fo, f1o
  * OK, names changed
* wrdlenccats1lenm1: this can be generalized to ( # P ) = ( ( # ( P concat <" A "> ) ) - 1 ) ). Also, the preferred variable used for words seems to be inconsistent, you should pick a convention and stick to it
  * generalized as proposed
  * fstwrdne preferred variable used for words adjusted P->W
  * fstwrdne0 preferred variable used for words adjusted P->W
  * ccatsymb preferred variable used for words adjusted B->V, S->A, T->B
* elovmpt2wrd, elovmptnn0wrd: seems a bit over-specific, there are a few options for generalization
  * will remain in AV's Mathbox
* swrdltnd, swrdnd: avoid multiple conjunct groups
  * to be clarified: do you mean better ( ph /\ ps ) -> th than ( ph -> ( ps -> th )?
* swrd0: what's up with the remark?
  * theorem generalized!
* addlenrevswrd: seems like the addends would make more sense the other way around
  * used in the context of cyclic shifts - here, this order is the natural one!
  * addlenswrd with reversed addends added
* disjxwrd: s/disjunct/disjoint/. Also it has an easy proof by invdisj
  * disjunct replaced by disjoint
  * proof shortened by invdisj
* swrd0swrd: could be proven as a special case of swrdswrd
  * proof shortened by swrdswrd
* swrdccat3a0: generalizes to ( (/) concat W ) = W
  * theorem removed (not used anyway)
* swrdccatin1d: deduction form should split up all the assumptions into different hypotheses
  * swrdccatin1d/swrdccatin2d/swrdccatin12d modified accordingly
* add2t1: sounds like it involves multiplication. Maybe one of add1add1 or add1p1 or p1p1?
  * renamed to add1p1
* ccat2s1cl, ...: is there a reason to not use the analogous theorems on <" X Y ">? The proofs are also much longer than if you just rewrote to this, maybe you weren't aware of them?
  * ccat2s1* theorems will remain in AV's Mathbox
* ccatw2s1ass: As above, I would want some specific advantage of being specific to singleton words. This should be ( ( W concat <" X "> ) concat <" Y "> ) = ( W concat <" X Y "> ), and there shouldn't be any other theorems about ( ( W concat <" X "> ) concat <" Y "> )
  * new theorem ccatw2s1ccatws2 added as proposed
  * other ccatw2s1* theorems except ccatw2s1ass will remain in AV's Mathbox
* ccatw2s1cl: reducible
  * how? Proof using ccatw2s1ccatws2 would be much longer
* ccat2s1fvw: I am concerned why this isn't a four line proof
  * 22 lines now instead of previously 41 ;-)
* df-csh: should be lowercase - cyclShift
  * renamed to cyclShift 
* df-lswrd: should be lowercase - lastS. Is last taken?
  * renamed to lastS 
  * although "last" seems not to be defined/used yet, I'd prefer lastS to indicate that it is the last "Symbol"
* cshwoor: could be merged with the previous theorem with -. N e. ( 0 ... ( # ` W ) ). We should try to minimize how much we discuss out of bounds behavior.
  * corresponding theorem cshwoofz added
* Also, couldn't df-csh be extended to ZZ by modular arithmetic? That would give it some more regular behavior, although you probably want to only talk about how it factors into substr when the input is between 0 and (# ` W)
  * added alternative definition as a remark in the comment
  * I'll check later if the definition can be extended without an impact on most of theorems about cyclical shifts. 
* lswrdn0: Really not a fan of the empty set assumption. This should just say that it is in V. Also, the name segment for this function is inconsistent. Is it ls or lswrd?
  * I think it is worth to mention this special case (empty set assumption) somewhere
  * ( lastS ` W ) e. V is shown in lswcl
  * will remain in AV's Mathbox
  * name segment should be "lsw" (adjusted everywhere)
* lswrd1: I think it would be nicer to just say ( lastS ` <" A "> ) = A. In fact you should have these theorems for all the explicit tuples, and a theorem about ( lastS ` ( S concat T ) ) = ( lastS ` T )
  * theorem for ( lastS ` <" A "> ) = A added (lsws1)
  * theorem for ( lastS ` ( S concat T ) ) = ( lastS ` T ) added (lswccatlsw)
  * theorems for other tuples should be added later when needed 
* lstccats1fst: this is bizzarely stated. The RHS should just be ( P ` 0 ). In fact it should be generalized to ( lastS ` ( S concat <" A "> ) ) = A, as a consequence of the other theorems I mentioned
  * proof shortened
  * will remain in AV's Mathbox
* swrdtrcfvl: the length of this proof is surprising. It should be swrd0fvls + swrd0len + algebra
  * proof shortened massively
* 2cshwcom: should follow from 2cshwmod + algebra
  * proof shortened by using 2cshwmod
* cshwsame: could be stated with A instead of ( W ` 0 ) in the hypothesis. Also, I think this is a good place to introduce the idiom ( ( 0 ..^ L ) X. { A } ) for the word "repeat A, L times"
  * cshwsame (and also cshwsame0, cshwssizesame) revised accordingly
  * new definition for word "repeat A, L times" (e.g. ` ( A repeat L ) ` )?
* cshwsdisj, ...: split up the deduction hypothesis
  * this would be a lot of work, because there are many theorems/lemmas depending on this ...
* cshwssize: it is called hash not size
  * "cshwssize" replaced by "cshwshash" in all labels
* cshwsexa: requires no assumptions
  * cshwsexa moved out of the block with the assumptions
* lstwrdne: duplicate of lswrdcl (also, more naming inconsistency)
  * removed
* lstwrdne0: I don't think we need it. The two assumptions can be merged to ``( # W ) e. NN, but that's just another way to say `W =/= (/)` which is the preferred idiom.
  * renamed to lswlgt0cl
  * will remain in AV's Mathbox
* 2swrd1eqwrdeq, 2swrd2eqwrdeq: again, avoid <" A "> = <" B ">
  * <" A "> = <" B "> replaced!
* ccat2s1lst: reducible
  * removed, can be replaced by lswccats1
* ccats1swrdeq: you could state this with ( lastS ` U ) instead of an existential
  * modified to use ( lastS ` U ) instead of an existential
* ccats1rev: These hypotheses are pretty crazy. If you want sufficient conditions for the implication, I would suggest ( lastS ` U ) = S, but then the conclusion is reducible to just U = ( W concat <" ( lastS ` U ) "> ).
  * will remain in AV's Mathbox

**Comment by avekens** [comment]
With my latest pull request, I finished moving the (new) definitions and theorems for words from my MathBox to the main part of set.mm. By this, steps 1.-3. in Gérards list are finished.

Regarding steps 4., many theorems about modulo and hash are already moved, because they are required by other theorems. Nevertheles, I will look for additional theorems to be moved in the next time.

After step 4. is done, I'll look what can be done regarding steps 5, and 6.

**Comment by david-a-wheeler** [comment]
Excellent! I want to see more things moved from the math boxes into the main body, and this certainly fits that bill.

**Comment by avekens** [comment]
On moving some theorems about the floor function from my Mathbox to the main part, I found that there are some theorems about the ("simulated") ceiling function in section 5.6.1. However, David provides a (corresponding!) definition of the ceiling function in his MathBox. I think this is a little bit confusing. Therefore I propose to move David's definition df-ceiling (calling it df-ceil?) and the corresponding theorems also into the main part, section 5.6.1, renaming this section "The floor and ceiling functions". Any objections?

**Comment by avekens** [comment]
Regarding section 5.6.2  "The modulo (remainder) operation": There is the following definition
  $c == $.  $( Equivalence symbol. $)
which seems to be used nowhere - can I delete it?
Maybe it could be useful to have such a symbol, but it should be defined and evaluated in sombody's MathBox first.

**Comment by david-a-wheeler** [comment]
I'd be delighted to see my ceiling function move into the main body, and feel free to rename things. In general, I'm delighted when stuff moves from my math box into the main body.

**Comment by nmegill** [comment]
@avekens  I think |^ would be acceptable.  I don't see any confusion with |^|, especially given that the first is usually used with function value ` and that they typically occur in different contexts.  -Norm

**Comment by avekens** [comment]
After finishing step 4. of Gérard's list, I think my job is done and this issue can be closed now.

The remaining steps of Gérard's list 
- 5: Finally, I would also add in the new paragraph « 5.7 Words over a set » the current section « 19.3.18.1 Operations on words » from Thierry’s Mathbox
- 6: And maybe also add in the current section « 5.7.2 Signum (sgn or sign) function « (to become section 5.8.2) Thierrys current section « 19.3.17 Signum (sgn or sign) function - misc additions » from theorem sgncl 25188 to theorem eluzmn 25203 (that do not use monois theory).
should be tracked by a separate issue. Additionally, someone else should do this (at best Thierry himself).
